controller/backing_image_controller.go

```diff
@@ -0,0 +1,605 @@
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Longhorn backing image %v out of the queue: %v", key, err)
+	bic.queue.Forget(key)
+}
+
+func getLoggerForBackingImage(logger logrus.FieldLogger, bi *longhorn.BackingImage) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"backingImageName": bi.Name,
+			"imageURL":         bi.Spec.ImageURL,
+		},
+	)
+}
+
+func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "fail to sync backing image for %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != bic.namespace {
+		return nil
+	}
+
+	backingImage, err := bic.ds.GetBackingImage(name)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			bic.logger.WithField("backingImage", name).WithError(err).Error("Failed to retrieve backing image from datastore")
+			return err
+		}
+		bic.logger.WithField("backingImage", name).Debug("Can't find backing image, may have been deleted")
+		return nil
+	}
+
+	log := getLoggerForBackingImage(bic.logger, backingImage)
+
+	if backingImage.Status.OwnerID != bic.controllerID {
+		if !bic.isResponsibleFor(backingImage) {
+			return nil
+		}
+		backingImage.Status.OwnerID = bic.controllerID
+		backingImage, err = bic.ds.UpdateBackingImageStatus(backingImage)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Backing Image got new owner %v", bic.controllerID)
+	}
+
+	if backingImage.Status.OwnerID != bic.controllerID {
+		// Not ours
+		return nil
+	}
+
+	if backingImage.DeletionTimestamp != nil {
+		pods, err := bic.ds.ListBackingImageRelatedPods(backingImage.Name)
+		if err != nil {
+			return err
+		}
+		// Rely on the foreground delete propagation to clean up all pods
+		if len(pods) == 0 {
+			log.Info("The data in all disks is removed, then the finalizer will be removed")
+			return bic.ds.RemoveFinalizerForBackingImage(backingImage)
+		}
+		return nil
+	}
+
+	existingBackingImage := backingImage.DeepCopy()
+	defer func() {
+		if err == nil && !reflect.DeepEqual(existingBackingImage.Status, backingImage.Status) {
+			_, err = bic.ds.UpdateBackingImageStatus(backingImage)
+		}
+		if apierrors.IsConflict(errors.Cause(err)) {
+			log.WithError(err).Debugf("Requeue %v due to conflict", key)
+			bic.enqueueBackingImage(backingImage)
+			err = nil
+		}
+	}()
+
+	if err := bic.syncBackingImageWithDownloadPods(backingImage); err != nil {
+		return err
+	}
+
+	if err := bic.updateDiskLastReferenceMap(backingImage); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getDownloadPodName(biName, diskUUID string) string {
+	return fmt.Sprintf("%s-download-%s", biName, diskUUID[:8])
+}
+
+func (bic *BackingImageController) syncBackingImageWithDownloadPods(backingImage *longhorn.BackingImage) (err error) {
```

Comment:
Might be useful to add a doc comment with all the state transitions like I do for the share-manager-controller, up to you though. example here: https://github.com/longhorn/longhorn-manager/blob/master/controller/share_manager_controller.go#L537-L543